### PR TITLE
Refactor title handling

### PR DESCRIPTION
**Pull request by benjiwheeler**
### Resolves

Should resolve https://github.com/LLK/scratch-gui/issues/3433 and https://github.com/LLK/scratch-desktop/issues/71 ; 

Note that it does not seem to resolve the related https://github.com/LLK/scratch-gui/issues/4759 .

### Proposed Changes

Repurpose TitledHOC so that it:
* no longer maintains projectTitle state
* no longer passes projectTitle down to children
* no longer sets `canEditTitle` to true on children
* has responsibility for setting redux projectTitle when parent passes a changed projectTitle prop (previously this functionality was in containers/gui.jsx)
* notices when the project-state changes to `isShowingWithoutId`, uses the translated default project name to set the redux projectTitle, and also calls the parent's`onUpdateProjectTitle` prop function

GUI's playground/standalone versions previously used TitledHOC, so they need to explicitly pass `canEditTitle` to children.

scratch-desktop will need to no longer wrap itself with TitledHOC, and instead maintain its own projectTitle state.

### Related

If this is merged, we will need to merge https://github.com/LLK/scratch-desktop/pull/74 as well so that scratch-desktop is in line with these changes.

### Reason for Changes

1) https://github.com/LLK/scratch-gui/issues/3433 and https://github.com/LLK/scratch-desktop/issues/71 were bugs
2) we want to localize the default project name

### CODE REVIEW

File: src/containers/gui.jsx [hunk 1]
```diff
@@ -126,17 +103,16 @@ GUI.propTypes = {
     isLoading: PropTypes.bool,
     isScratchDesktop: PropTypes.bool,
     isShowingProject: PropTypes.bool,
+    isShowingWithoutId: PropTypes.bool,
```
Comment (rschamp) [inline src/containers/gui.jsx]: Is this added intentionally? It looks unrelated.

File: src/lib/titled-hoc.jsx [hunk 2]
```diff
@@ -12,26 +26,89 @@ const TitledHOC = function (WrappedComponent) {
...
+            if (this.props.projectTitle !== prevProps.projectTitle) {
+                this.setReduxTitle(this.props.projectTitle);
+            }
+            if (this.props.isShowingWithoutId && !prevProps.isShowingWithoutId) {
+                const defaultProjectTitle = this.titleWithDefault();
+                this.setReduxTitle(defaultProjectTitle);
+                this.props.onUpdateProjectTitle(defaultProjectTitle);
+            }
+        }
+        titleWithDefault (title) {
```
Comment (rschamp) [inline src/lib/titled-hoc.jsx]: Why is this a function that takes a `title` argument? It looks like it's only called in one place, without an argument.

File: src/lib/titled-hoc.jsx [hunk 3]
```diff
@@ -12,26 +26,89 @@ const TitledHOC = function (WrappedComponent) {
...
+                // redux value
+                projectTitle,
+                /* eslint-enable no-unused-vars */
+                ...componentProps
+            } = this.props;
             return (
                 <WrappedComponent
-                    canEditTitle
-                    projectTitle={this.state.projectTitle}
                     onUpdateProjectTitle={this.handleUpdateProjectTitle}
```
Comment (rschamp) [inline src/lib/titled-hoc.jsx]: If we expect children to receive the title from redux, I think we should we also expect them to set it through redux.

File: src/containers/gui.jsx (hunk 1 above)
Comment (benjiwheeler) [reply src/containers/gui.jsx]: Removing...

File: src/lib/titled-hoc.jsx (hunk 3 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: I made this change, now testing it...

File: src/lib/titled-hoc.jsx (hunk 2 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: I think I meant to refactor setReduxTitle further... doing this now

File: src/containers/gui.jsx (hunk 1 above)
Comment (benjiwheeler) [reply src/containers/gui.jsx]: Done

File: src/lib/titled-hoc.jsx (hunk 2 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: Done

File: src/lib/titled-hoc.jsx (hunk 3 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: Tested well

File: src/lib/titled-hoc.jsx [hunk 4]
```diff
@@ -1,37 +1,97 @@
...
-                projectTitle: null
-            };
+        componentDidMount () {
+            this.props.updateReduxProjectTitle(this.titleWithDefault(this.props.projectTitle));
+        }
+        componentDidUpdate (prevProps) {
+            if (this.props.projectTitle !== prevProps.projectTitle) {
+                this.props.updateReduxProjectTitle(this.titleWithDefault(this.props.projectTitle));
+            }
+            // if the projectTitle hasn't changed, but the reduxProjectTitle
```
Comment (benjiwheeler) [inline src/lib/titled-hoc.jsx]: Note this addition... here is where the change to redux title gets echoed up to the enclosing context (i.e., www)

File: src/components/menu-bar/project-title-input.jsx [hunk 5]
```diff
@@ -19,47 +19,37 @@ const messages = defineMessages({
     }
 });
 
-class ProjectTitleInput extends React.Component {
```
Comment (benjiwheeler) [inline src/components/menu-bar/project-title-input.jsx]: I should change this back to how it was before

File: src/lib/titled-hoc.jsx [hunk 6]
```diff
@@ -1,37 +1,97 @@
...
-        constructor (props) {
-            super(props);
-            bindAll(this, [
-                'handleUpdateProjectTitle'
-            ]);
-            this.state = {
-                projectTitle: null
-            };
+        componentDidMount () {
+            this.props.updateReduxProjectTitle(this.titleWithDefault(this.props.projectTitle));
```
Comment (benjiwheeler) [inline src/lib/titled-hoc.jsx]: can combine these calls

File: src/components/menu-bar/project-title-input.jsx (hunk 5 above)
Comment (rschamp) [reply src/components/menu-bar/project-title-input.jsx]: I think this is fine to do since your handler is coming from Redux now. I just think the name of the prop should change.

File: src/playground/player.jsx [hunk 7]
```diff
@@ -24,6 +23,7 @@ const Player = ({isPlayerOnly, onSeeInside, projectId}) => (
     <Box className={classNames(isPlayerOnly ? styles.stageOnly : styles.editor)}>
         {isPlayerOnly && <button onClick={onSeeInside}>{'See inside'}</button>}
         <GUI
+            canEditTitle
```
Comment (rschamp) [inline src/playground/player.jsx]: Curious why this is necessary to add to the playgrounds.

Comment (rschamp) [reply src/playground/player.jsx]: Nevermind I see now!

File: src/components/menu-bar/project-title-input.jsx (hunk 5 above)
Comment (benjiwheeler) [reply src/components/menu-bar/project-title-input.jsx]: Doing this...

File: src/components/menu-bar/project-title-input.jsx [hunk 8]
```diff
@@ -19,47 +19,37 @@ const messages = defineMessages({
...
+        tabIndex="0"
+        type="text"
+        value={projectTitle}
+        onSubmit={handleUpdateReduxProjectTitle}
+    />
+);
 
 ProjectTitleInput.propTypes = {
     className: PropTypes.string,
+    handleUpdateReduxProjectTitle: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/menu-bar/project-title-input.jsx]: What I'm used to the linter insisting on is that this form be followed:

```
onXXXX={this.handleXXXX}
```

In my experience it doesn't like
```
onXXXX={this.onXXXX}
```

but it's ok with
```
onXXXX={this.props.whateverNameIsOK}
```

File: src/containers/sb-file-uploader.jsx [hunk 9]
```diff
@@ -179,9 +180,9 @@ SBFileUploader.propTypes = {
     loadingState: PropTypes.oneOf(LoadingStates),
     onLoadingFinished: PropTypes.func,
     onLoadingStarted: PropTypes.func,
-    onUpdateProjectTitle: PropTypes.func,
     projectChanged: PropTypes.bool,
     requestProjectUpload: PropTypes.func,
+    updateReduxProjectTitle: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/containers/sb-file-uploader.jsx]: Changing...

File: src/lib/titled-hoc.jsx (hunk 6 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: Changing...

File: src/components/menu-bar/project-title-input.jsx (hunk 5 above)
Comment (benjiwheeler) [reply src/components/menu-bar/project-title-input.jsx]: Done

File: src/lib/titled-hoc.jsx (hunk 6 above)
Comment (benjiwheeler) [reply src/lib/titled-hoc.jsx]: Done

File: src/containers/sb-file-uploader.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/containers/sb-file-uploader.jsx]: Done
